File: app/src/main/java/com/a60n1/ejashkojme/adapter/CommentAdapter.java
package com.a60n1.ejashkojme.adapter;

import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.a60n1.ejashkojme.MainActivity;
import com.a60n1.ejashkojme.ProfileActivity;
import com.a60n1.ejashkojme.R;
import com.a60n1.ejashkojme.models.Comment;
import com.a60n1.ejashkojme.models.User;
import com.a60n1.ejashkojme.viewholder.CommentViewHolder;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Callback;
import com.squareup.picasso.NetworkPolicy;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class CommentAdapter extends RecyclerView.Adapter<CommentViewHolder> {
    private static final String TAG = "CommentAdapter";
    private Context mContext;
    private MainActivity mainActivity;
    private DatabaseReference mDatabaseReference;
    private ChildEventListener mChildEventListener;

    private List<String> mCommentIds = new ArrayList<>();
    private List<Comment> mComments = new ArrayList<>();

    public CommentAdapter(final Context context, DatabaseReference ref) {
        mContext = context;
        mDatabaseReference = ref;
        mainActivity = (MainActivity) mContext;
        // Create child event listener
        ChildEventListener childEventListener = new ChildEventListener() {
            @Override
            public void onChildAdded(DataSnapshot dataSnapshot, String previousChildName) {
                Log.d(TAG, "onChildAdded:" + dataSnapshot.getKey());

                // A new comment has been added, add it to the displayed list
                Comment comment = dataSnapshot.getValue(Comment.class);

                // Update RecyclerView
                mCommentIds.add(dataSnapshot.getKey());
                mComments.add(comment);
                notifyItemInserted(mComments.size() - 1);
            }

            @Override
            public void onChildChanged(DataSnapshot dataSnapshot, String previousChildName) {
                Log.d(TAG, "onChildChanged:" + dataSnapshot.getKey());

                // A comment has changed, use the key to determine if we are displaying this
                // comment and if so displayed the changed comment.
                Comment newComment = dataSnapshot.getValue(Comment.class);
                String commentKey = dataSnapshot.getKey();

                int commentIndex = mCommentIds.indexOf(commentKey);
                if (commentIndex > -1) {
                    // Replace with the new data
                    mComments.set(commentIndex, newComment);

                    // Update the RecyclerView
                    notifyItemChanged(commentIndex);
                } else {
                    Log.w(TAG, "onChildChanged:unknown_child:" + commentKey);
                }
            }

            @Override
            public void onChildRemoved(DataSnapshot dataSnapshot) {
                Log.d(TAG, "onChildRemoved:" + dataSnapshot.getKey());

                // A comment has changed, use the key to determine if we are displaying this
                // comment and if so remove it.
                String commentKey = dataSnapshot.getKey();

                int commentIndex = mCommentIds.indexOf(commentKey);
                if (commentIndex > -1) {
                    // Remove data from the list
                    mCommentIds.remove(commentIndex);
                    mComments.remove(commentIndex);

                    // Update the RecyclerView
                    notifyItemRemoved(commentIndex);
                } else {
                    Log.w(TAG, "onChildRemoved:unknown_child:" + commentKey);
                }
            }

            @Override
            public void onChildMoved(DataSnapshot dataSnapshot, String previousChildName) {
                Log.d(TAG, "onChildMoved:" + dataSnapshot.getKey());

                // A comment has changed position, use the key to determine if we are
                // displaying this comment and if so move it.
                Comment movedComment = dataSnapshot.getValue(Comment.class);
                String commentKey = dataSnapshot.getKey();
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
                Log.w(TAG, "postComments:onCancelled", databaseError.toException());
                Toast.makeText(mContext, "Failed to load comments.",
                        Toast.LENGTH_SHORT).show();
            }
        };
        ref.addChildEventListener(childEventListener);

        // Store reference to listener so it can be removed on app stop
        mChildEventListener = childEventListener;
    }

    @NonNull
    @Override
    public CommentViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(mContext);
        View view = inflater.inflate(R.layout.item_comment, parent, false);
        return new CommentViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final CommentViewHolder holder, int position) {
        final Comment comment = mComments.get(position);
        holder.authorView.setText(comment.author);
        holder.bodyView.setText(comment.text);
        holder.authorView.setOnClickListener(view -> {
            PopupMenu popup = new PopupMenu(mContext, holder.authorView);
            popup.inflate(R.menu.menu_user_action);
            popup.setOnMenuItemClickListener(item -> {
                if (item.getItemId() == R.id.profile_action) {
                    Intent intent = new Intent(mContext, ProfileActivity.class);
                    intent.putExtra("user_id", comment.uid);
                    mContext.startActivity(intent);
                }
                return false;
            });
            popup.show();
        });
        String uid = comment.uid;
        DatabaseReference userDatabase = FirebaseDatabase.getInstance().getReference()
                .child("users").child(uid);
        userDatabase.keepSynced(true);
        ValueEventListener userListener = new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                User user = dataSnapshot.getValue(User.class);
                if (user == null) {
                    return;
                }
                final String thumb_image = user.thumb_image;
                if (!thumb_image.equals("default")) {
                    Picasso.get().load(thumb_image).networkPolicy(NetworkPolicy.OFFLINE)
                            .placeholder(R.drawable.default_avatar).into(holder.authorAvatar, new Callback() {
                        @Override
                        public void onSuccess() {

                        }

                        @Override
                        public void onError(Exception e) {
                            Picasso.get().load(thumb_image).placeholder(R.drawable.default_avatar).into(holder.authorAvatar);
                        }
                    });
                } else {
                    Picasso.get().load(thumb_image).placeholder(R.drawable.default_avatar).into(holder.authorAvatar);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
            }
        };
        userDatabase.addValueEventListener(userListener);
    }

    @Override
    public int getItemCount() {
        return mComments.size();
    }

    public void cleanupListener() {
        if (mChildEventListener != null) {
            mDatabaseReference.removeEventListener(mChildEventListener);
        }
    }

}

File: app/src/main/java/com/a60n1/ejashkojme/adapter/CommentAdapter.kt
package com.a60n1.ejashkojme.adapter

import android.content.Context
import android.content.Intent
import android.util.Log
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.ViewGroup
import android.widget.PopupMenu
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.a60n1.ejashkojme.MainActivity
import com.a60n1.ejashkojme.ProfileActivity
import com.a60n1.ejashkojme.R
import com.a60n1.ejashkojme.models.Comment
import com.a60n1.ejashkojme.models.User
import com.a60n1.ejashkojme.viewholder.CommentViewHolder
import com.google.firebase.database.*
import com.squareup.picasso.Callback
import com.squareup.picasso.NetworkPolicy
import com.squareup.picasso.Picasso
import java.util.*

class CommentAdapter(private val mContext: Context, private val mDatabaseReference: DatabaseReference)
    : RecyclerView.Adapter<CommentViewHolder>() {
    @Suppress("JoinDeclarationAndAssignment")
    private val mainActivity: MainActivity
    private val mChildEventListener: ChildEventListener?
    private val mCommentIds: MutableList<String?> = ArrayList()
    private val mComments: MutableList<Comment> = ArrayList()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CommentViewHolder {
        val inflater = LayoutInflater.from(mContext)
        val view = inflater.inflate(R.layout.item_comment, parent, false)
        return CommentViewHolder(view)
    }

    override fun onBindViewHolder(holder: CommentViewHolder, position: Int) {
        val comment = mComments[position]
        holder.authorView.text = comment.author
        holder.bodyView.text = comment.text
        holder.authorView.setOnClickListener {
            val popup = PopupMenu(mContext, holder.authorView)
            popup.inflate(R.menu.menu_user_action)
            popup.setOnMenuItemClickListener { item: MenuItem ->
                if (item.itemId == R.id.profile_action) {
                    val intent = Intent(mContext, ProfileActivity::class.java)
                    intent.putExtra("user_id", comment.uid)
                    mContext.startActivity(intent)
                }
                false
            }
            popup.show()
        }
        val uid = comment.uid
        val userDatabase = FirebaseDatabase.getInstance().reference
                .child("users").child(uid!!)
        userDatabase.keepSynced(true)
        val userListener: ValueEventListener = object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val user = dataSnapshot.getValue(User::class.java) ?: return
                val thumbImage = user.thumbImage
                if (thumbImage != "default")
                    Picasso.get().load(thumbImage).networkPolicy(NetworkPolicy.OFFLINE)
                            .placeholder(R.drawable.default_avatar).into(holder.authorAvatar, object : Callback {
                                override fun onSuccess() {}
                                override fun onError(e: Exception) {
                                    Picasso.get().load(thumbImage).placeholder(R.drawable.default_avatar).into(holder.authorAvatar)
                                }
                            })
                else
                    Picasso.get().load(thumbImage).placeholder(R.drawable.default_avatar).into(holder.authorAvatar)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        }
        userDatabase.addValueEventListener(userListener)
    }

    override fun getItemCount(): Int {
        return mComments.size
    }

    fun cleanupListener() {
        if (mChildEventListener != null)
            mDatabaseReference.removeEventListener(mChildEventListener)
    }

    companion object {
        private const val TAG = "460N_DEV_CommentAdapter"
    }

    init {
        mainActivity = mContext as MainActivity
        // Krijojme ChildEventListener qe prej events
        val childEventListener: ChildEventListener = object : ChildEventListener {
            override fun onChildAdded(dataSnapshot: DataSnapshot, previousChildName: String?) {
                Log.d(TAG, "onChildAdded:" + dataSnapshot.key)
                // Kur ndodh shtimi i komentit
                val comment = dataSnapshot.getValue(Comment::class.java)!!
                // Shtohet ne RecyclerView
                mCommentIds.add(dataSnapshot.key)
                mComments.add(comment)
                notifyItemInserted(mComments.size - 1)
            }

            override fun onChildChanged(dataSnapshot: DataSnapshot, previousChildName: String?) {
                Log.d(TAG, "onChildChanged:" + dataSnapshot.key)
                // Koment ka ndryshu
                // per te ne ardhmen kur edit comment shtohet si feature
                val newComment = dataSnapshot.getValue(Comment::class.java)!!
                val commentKey = dataSnapshot.key
                val commentIndex = mCommentIds.indexOf(commentKey)
                if (commentIndex > -1) { // zevendesimi
                    mComments[commentIndex] = newComment
                    // update UI
                    notifyItemChanged(commentIndex)
                } else
                    Log.w(TAG, "onChildChanged:unknown_child:$commentKey")
            }

            override fun onChildRemoved(dataSnapshot: DataSnapshot) {
                Log.d(TAG, "onChildRemoved:" + dataSnapshot.key)
                // Largimi i komenteve
                // per ne te ardhmen kur delete comment shtohet si feature
                val commentKey = dataSnapshot.key
                val commentIndex = mCommentIds.indexOf(commentKey)
                if (commentIndex > -1) { // largimi prej listes tkomenteve
                    mCommentIds.removeAt(commentIndex)
                    mComments.removeAt(commentIndex)
                    // update UI
                    notifyItemRemoved(commentIndex)
                } else
                    Log.w(TAG, "onChildRemoved:unknown_child:$commentKey")
            }

            @Suppress("UNUSED_VARIABLE")
            override fun onChildMoved(dataSnapshot: DataSnapshot, previousChildName: String?) {
                Log.d(TAG, "onChildMoved:" + dataSnapshot.key)
                // Funksion i nevojshem
                // nuk perdoret, vetem template
                val movedComment = dataSnapshot.getValue(Comment::class.java)!!
                val commentKey = dataSnapshot.key
            }

            override fun onCancelled(databaseError: DatabaseError) {
                Log.w(TAG, "postComments:onCancelled", databaseError.toException())
                // deshtim i leximit per qfaredo arsye
                Toast.makeText(mContext, "Failed to load comments.",
                        Toast.LENGTH_SHORT).show()
            }
        }
        mDatabaseReference.addChildEventListener(childEventListener)
        // shtimi ne reference qe te ndalet kur aplikacioni te ndalet
        mChildEventListener = childEventListener
    }
}